### PR TITLE
[BugFix][Kernel] Fix and optimize sampling mask packing on Ascend

### PR DESCRIPTION
**Pull request by Liamup777**
### What this PR does / why we need it?

The sampling-mask kernel can overflow UB on Triton-Ascend when
`logits.stride(1) > 1`. For a batch-32, vocabulary-151936, stride-2
input, the original BLOCK_SIZE=8192 kernel fails compilation with:

`ub overflow, requires 12583424 bits while 1572864 bits available`

The implementation also has two Ascend-specific lowering issues:

- reducing the boolean keep mask does not produce the required finite-logit
  count;
- per-element variable shifts and the `[512, 8]` axis-1 reduction lower to
  scalar-heavy execution.

This PR adds an Ascend implementation for the three-field sampling-mask
bitmask API. It:

- makes only strided vocabulary inputs contiguous, without modifying the
  caller's tensor;
- casts the keep mask to int32 before reduction;
- uses BLOCK_SIZE=4096 to remain within the NPU UB limit;
- transposes the packing layout from `[512, 8]` to `[8, 512]`, multiplies by
  compile-time bit weights, and reduces axis 0;
- directly replaces `SamplingMaskTensors.from_logits` for the target
  three-field sampling-mask API;
- adds operator documentation and registers the NPU regression test in the
  A2 single-card nightly matrix.

The following msprof results use an Ascend 910, `npu:0`, 10 profiling
rounds, and `MemoryDetail,Occupancy,PipeUtilization,Roofline`. Values are
selected kernel task durations and exclude Python wrapper time and the
preceding contiguous copy.

| Case | Baseline | Optimized | Speedup | Reduction |
| --- | ---: | ---: | ---: | ---: |
| `b1_v17_active1_colstride1` | 123.330 us | 5.640 us | 21.87x | 95.43% |
| `b8_v8191_active1_colstride1` | 244.573 us | 6.960 us | 35.14x | 97.15% |
| `b8_v8192_active1_colstride1` | 244.327 us | 6.680 us | 36.58x | 97.27% |
| `b8_v8193_active1_colstride1` | 365.349 us | 8.940 us | 40.87x | 97.55% |
| `b8_v8193_active0_colstride1` | 365.357 us | 8.780 us | 41.61x | 97.60% |
| `b32_v151936_active1_colstride2` | 4604.258 us | 84.758 us | 54.32x | 98.16% |
| `b128_v151936_active1_colstride1` | 18403.968 us | 327.093 us | 56.27x | 98.22% |

The original stride-2 BLOCK_SIZE=8192 implementation has no runtime
duration because it fails compilation with UB overflow. The table uses
the first compilable correctness-fixed implementation with the original
variable-shift packing as the performance baseline.

TTIR/NPU IR inspection confirms that the optimized packing lowers to
vector transpose, multiply, and reduction operations; the previous
tensor-wide scalar shift loop is absent.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Test environment:

- Ascend 910
- Python 3.12.13
- Torch 2.10.0
- torch-npu 2.10.0.post4
- Triton-Ascend 3.2.0
- CANN 9.1.0

The NPU regression test passed in both supported dependency lanes.

The regression test verifies:

- non-contiguous `logits` with `stride(1) == 2`;
- exact packed masks and finite-logit counts;
- active and inactive request rows;
- the caller's logits view remains non-contiguous after the call.

The regression is registered as the A2 single-card nightly case
`sampling-mask`. The nightly matrix resolver was run locally and produced
the expected test path and `linux-aarch64-a2b3-1` runner.

Seven profiling cases covering small, tile-boundary, inactive, strided,
large-vocabulary, and large-batch inputs passed correctness checks. An
additional patterned vocabulary-size-17 case passed exact bit-order,
tail-byte, and count checks.

The following repository checks passed:

- `bash format.sh`
- `python -m compileall`
- `git diff --check`
- nightly matrix resolution for `a2.single_node.test_config`
- `SHELLCHECK_OPTS="--exclude=SC2046,SC2006,SC2086" pre-commit run
  --all-files --hook-stage manual --show-diff-on-failure`

- vLLM main: https://github.com/vllm-project/vllm/commit/84030bbe3d74d99bad477a3d2e37a973ccd8865c

### CODE REVIEW

**Review by Liamup777 — COMMENTED** [review]
Additional context about the three-field/four-field compatibility and
Triton-Ascend support:

- vLLM v0.29.0 (`98dff2a81d`) uses the three-field
  `SamplingMaskTensors(packed_mask, counts, vocab_size)` interface.
- Later vLLM main (`84030bbe3`) uses the four-field
  `SamplingMaskTensors(token_ids, packed_mask, counts, vocab_size)` interface.
  Its compact kernel additionally generates `token_ids` through a
  `tl.cumsum`-based dynamic scatter.

The currently supported Triton-Ascend 3.2.2 can compile and run the Ascend
packing kernel, but the compact token-ID path may hang while lowering the
dynamic scatter. Therefore, this PR routes both interfaces through
`_pack_sampling_mask_kernel_ascend`.

For the four-field interface, the wrapper returns a zero-width `token_ids`
tensor. The existing upstream `tolists()` implementation then detects that the
row count exceeds the compact width and reconstructs the exact token IDs from
`packed_mask`.

This fallback preserves sampling correctness. Its tradeoff is that the
four-field path temporarily loses the compact token-ID transfer optimization
and may perform more D2H transfer and CPU bitmask decoding. A warning is emitted
once per process when this fallback is selected.

Validation on Ascend with Triton-Ascend 3.2.2:

- vLLM v0.29.0 three-field API: 4 tests passed.
- pinned vLLM main `84030bbe3` four-field API: 4 tests passed.

In local A3 testing, Triton-Ascend `3.6.0+dev20260920` completed both the
minimal dynamic-scatter reproducer and the real compact sampling-mask case.
However, vLLM-Ascend currently pins Triton-Ascend 3.2.2, so this PR cannot rely
on that unreleased/nightly behavior.

The packed-mask fallback can be removed and the compact token-ID path restored
after a Triton-Ascend release containing the fix becomes the project's minimum
supported version.

File: vllm_ascend/ops/triton/v2/sample/sampling_mask.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            mask=valid,
+            other=-float("inf"),
+        )
+        keep = (logits > -float("inf")) & (logits < float("inf")) & is_active
+        keep_i32 = keep.to(tl.int32)
+        # Triton-Ascend reduces an i1 tensor as a boolean. Cast before the sum
+        # so counts contains the number of finite logits, not the tile count.
+        count += tl.sum(keep_i32)
+
+        keep = tl.trans(tl.reshape(keep_i32, (BLOCK_SIZE // 8, 8)))
```
Comment (AuroraEmiya) [inline vllm_ascend/ops/triton/v2/sample/sampling_mask.py]: be specific

**Review by AuroraEmiya — APPROVED** [review]
LGTM

File: .github/workflows/configs/nightly_config.yaml [hunk 2]
```diff
@@ -11,6 +11,10 @@
 a2:
   single_node:
     test_config:
+      # Pytest-driven single-operator tests
```
Comment (ZT-AIA) [inline .github/workflows/configs/nightly_config.yaml:14]: Please delete the modifications here. That's all.
